src/lumps/textureLibrary.ts

```diff
@@ -0,0 +1,135 @@
+import {WADFlat} from "@src/lumps/doom/flat";
+import {WADTexture, WADTextures} from "@src/lumps/doom/textures";
+import {WADFileList} from "@src/wad/fileList";
+import {WADLump} from "@src/wad/lump";
+
+// Flats and walls use different texture sets.
+export enum TextureSet {
+    Walls,
+    Flats,
+}
+
+export class TextureLibrary {
+    // The textures themselves
+    protected textures: {
+        [TextureSet.Walls]: {[name: string]: WADTexture | null};
+        [TextureSet.Flats]: {[name: string]: WADFlat | null};
+    };
+    // And whether or not each one is transparent
+    protected transparent: {
+        [TextureSet.Walls]: {[name: string]: boolean | null};
+        [TextureSet.Flats]: {[name: string]: boolean | null};
+    };
+    // RGBA pixel buffers created from each texture
+    protected rgba: {
+        [TextureSet.Walls]: {[name: string]: Buffer | null};
+        [TextureSet.Flats]: {[name: string]: Buffer | null};
+    };
+    public fileList: WADFileList;
+    constructor(fileList: WADFileList){
+        this.fileList = fileList;
+        this.textures = {
+            [TextureSet.Walls]: {},
+            [TextureSet.Flats]: {},
+        };
+        this.transparent = {
+            [TextureSet.Walls]: {},
+            [TextureSet.Flats]: {},
+        };
+        this.rgba = {
+            [TextureSet.Walls]: {},
+            [TextureSet.Flats]: {},
+        };
+    }
+    protected isFlatMarker(lump: WADLump): boolean {
+        const name = lump.name;
+        return (
+            name === "F1_START" || name === "F2_START" || name === "F3_START" ||
+            name === "F1_END" || name === "F2_END" || name === "F3_END" ||
+            lump.dataLength === 0
+        );
```

Comment:
Please check out `lumpType.ts` L247 (in master anyway, it might be different in this branch)

One of these two checks must be wrong. I am _pretty_ sure that the one in lumpType is correct but it is very possible that we should be checking for all of the combined marker names and not only those - I _think_ F1_START/F1_END etc. can work in IWADs and not PWADs, but F_START/F_END and FF_START/FF_END will work in both. Regardless of whatever is the correct way to check, I would suggest placing a single source of truth for what flat markers can be named as static attributes of `WADFlat` in `flat.ts`, e.g.

``` ts
static readonly StartMarkerNames: string[] = ["F_START", "FF_START", "F1_START", "F2_START", "F3_START"];
static readonly EndMarkerNames: string[] = ["F_END", "FF_END", "F1_END", "F2_END", "F3_END"];
```